core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogBomScanner.java

```diff
@@ -155,9 +181,42 @@ public void removedBundle(Bundle bundle, BundleEvent bundleEvent, Iterable<? ext
                 LOG.debug("No BOM found in {} {} {}", bundleIds(bundle));
             }
 
+            if (!passesWhiteAndBlacklists(bundle)) {
+                removeAnyApplications(catalogItems);
+            }
+
             return catalogItems;
         }
 
+        private void removeAnyApplications(Iterable<? extends CatalogItem<?, ?>> catalogItems) {
+
+            final Iterator<? extends CatalogItem<?, ?>> cit = catalogItems.iterator();
+            while (cit.hasNext()) {
+                final CatalogItem<?, ?> item = cit.next();
+                if (TEMPLATE.equals(item.getCatalogItemType())) {
+                    removeFromCatalog(item);
+                    cit.remove();
```

Comment:
(minor) No need for an iterator & removing the items from `catalogItems`. `removeFromCatalog` does all that's needed.